### PR TITLE
Move to serverless client

### PR DESCRIPTION
**Pull request by izellevy**
## Problem

This PR replaces the current client with the new client that has support for serverless.

## Solution

Currently, we integrate with the dev release of pinecone-client.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

Made the necessary changes to be compatible with the new client.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -10,12 +10,12 @@ packages = [{include = "canopy", from = "src"},
             {include = "canopy_server", from = "src"},]
 
 [tool.poetry.dependencies]
-python = "^3.9"
-pinecone-client = "^2.2.2"
+python = ">=3.9,<3.13"
+pinecone-client = "3.0.0.dev4"
+pandas = "^2.0.0"
```
[the remark this answers is not in the document]
Comment (igiloh-pinecone) [reply pyproject.toml:15]: I agree with Roy. Unless we have a really good reason to restrict the version of some dependency - we should try to remain as loose as possible.

Any stricter versioning may create more headache for users who are installing canopy in a live production environments.

File: src/canopy/knowledge_base/knowledge_base.py [hunk 2]
```diff
@@ -102,13 +66,25 @@ class KnowledgeBase(BaseKnowledgeBase):
         'reranker': TransparentReranker
     }
 
+    class _VectorSetMetadata:
+        schema = [
+            ("id", False, None),
+            ("values", False, None),
+            ("sparse_values", True, None),
+            ("metadata", True, None),
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:74]: No I just took it from pinecone-datasets. We remove the dataset dependency that's why I copied the schema and related functionality.

File: src/canopy/knowledge_base/knowledge_base.py [hunk 3]
```diff
@@ -191,21 +168,44 @@ def __init__(self,
         # `create_canopy_index()`
         self._index: Optional[Index] = None
 
+    def list_canopy_indexes(self) -> List[str]:
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:171]: Done

File: src/canopy/knowledge_base/knowledge_base.py [hunk 4]
```diff
@@ -191,21 +168,44 @@ def __init__(self,
...
+        List all Canopy indexes in the current Pinecone account.
+
+        Returns:
+            A list of Canopy index names.
+        """
+        index_names = [_['name'] for _ in self._pc.list_indexes().index_list['indexes']]
+
+        return [index for index in index_names if index.startswith(INDEX_NAME_PREFIX)]
+
+    def connect_to_pinecone(self):
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:182]: Done

File: src/canopy/knowledge_base/knowledge_base.py [hunk 5]
```diff
@@ -543,21 +530,7 @@ def upsert(self,
...
-                                           dense_model=DenseModelMetadata(
-                                               name=encoder_name,
-                                               dimension=self._encoder.dimension),
-                                           queries=0)
-
-        dataset = Dataset.from_pandas(
-            pd.DataFrame.from_records([c.to_db_record() for c in encoded_chunks]),
-            metadata=dataset_metadata
-        )
+        encoded_df = pd.DataFrame.from_records([c.to_db_record() for c in encoded_chunks])
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:533]: Done

File: src/canopy_cli/cli.py [hunk 6]
```diff
@@ -81,7 +80,7 @@ def wait_for_server(chat_server_url: str):
 
 def validate_pinecone_connection():
     try:
-        connect_to_pinecone()
+        KnowledgeBase(index_name="test-conn").connect_to_pinecone()
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy_cli/cli.py:83]: Removed

File: src/canopy/knowledge_base/knowledge_base.py [hunk 7]
```diff
@@ -37,39 +32,56 @@
 DELETE_STARTER_CHUNKS_PER_DOC = 32
 
 
-def connect_to_pinecone():
+@lru_cache(maxsize=1)
+def _get_global_client() -> Pinecone:
+    return Pinecone()
+
+
+def connect_to_pinecone(pinecone_client: Pinecone = None):
```
Comment (miararoy) [inline src/canopy/knowledge_base/knowledge_base.py:40]: Consider changing function name to "check_pinecone_client_connection" this method is not doing "init" anymore

**Review by miararoy — COMMENTED** [review]
General review, pending @igiloh-pinecone review

File: src/canopy/knowledge_base/knowledge_base.py [hunk 8]
```diff
@@ -37,39 +32,56 @@
...
     Returns:
         A list of Canopy index names.
     """
+    pinecone_client = pinecone_client or _get_global_client()
 
-    connect_to_pinecone()
-    return [index for index in list_indexes() if index.startswith(INDEX_NAME_PREFIX)]
+    connect_to_pinecone(pinecone_client)
 
+    indexes = pinecone_client.list_indexes()
```
[the remark this answers is not in the document]
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:78]: Removed the connect you are right

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -17,6 +17,10 @@ on:
           - 'patch' # bug fixes
           - 'minor' # new features, backwards compatible
           - 'major' # breaking changes
+          - 'prepatch' # bumps the patch and moves to a prerelease (1.0.2 -> 1.0.3a0)
+          - 'preminor' # bumps the minor and moves to a prerelease (1.0.2 -> 1.1.0a0)
+          - 'premajor' # bumps the major and moves to a prerelease (1.0.2 -> 2.0.0a0)
+          - 'prerelease' # bumps the prerelease version (1.0.3a0 -> 1.0.3a1)
```
Comment (igiloh-pinecone) [inline .github/workflows/release.yml:23]: Theoretically, you also need to pass `prerelease: True` to the "Create GH release" step.  
(Not critical, I guess)

File: pyproject.toml [hunk 10]
```diff
@@ -5,17 +5,16 @@ description = "Retrieval Augmented Generation (RAG) framework and context engine
 authors = ["Relevance Team <relevance@pinecone.io>"]
 readme = "README.md"
 license = "Apache-2.0"
-packages = [{include = "canopy", from = "src"},
-            {include = "canopy_cli", from = "src"},
-            {include = "canopy_server", from = "src"},]
+packages = [{ include = "canopy", from = "src" },
+    { include = "canopy_cli", from = "src" },
+    { include = "canopy_server", from = "src" }, ]
```
Comment (igiloh-pinecone) [inline pyproject.toml:10]: Black!!!! 😠 

(Just kidding, don't change that back...)

File: src/canopy/knowledge_base/knowledge_base.py [hunk 11]
```diff
@@ -37,39 +32,40 @@
...
-    """
-    try:
-        pinecone_init()
-        pinecone_whoami()
-    except Exception as e:
-        raise RuntimeError("Failed to connect to Pinecone. "
-                           "Please check your credentials and try again") from e
+@lru_cache(maxsize=1)
+def _get_global_client() -> Pinecone:
+    return Pinecone()
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:37]: I'm not 100% sure this is multi-processing safe, but let's worry about that in the future

File: src/canopy/knowledge_base/knowledge_base.py [hunk 12]
```diff
@@ -109,6 +105,7 @@ def __init__(self,
                  chunker: Optional[Chunker] = None,
                  reranker: Optional[Reranker] = None,
                  default_top_k: int = 5,
+                 pinecone_client: Optional[Pinecone] = None
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:108]: This should probably be the 3rd argument, right after `index_name` (just for readability - the order doesn't really matter since we don't allow unnamed args anyway)

File: src/canopy/knowledge_base/knowledge_base.py [hunk 13]
```diff
@@ -259,7 +262,10 @@ def verify_index_connection(self) -> None:
             ) from e
 
     def create_canopy_index(self,
-                            indexed_fields: Optional[List[str]] = None,
+                            spec: Union[Dict, ServerlessSpec, PodSpec] = ServerlessSpec(
+                                cloud='aws',
+                                region='us-west-2'
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:267]: We should probably make these arguments of `create_caopy_index()` (with strong defaults.

File: src/canopy/knowledge_base/knowledge_base.py [hunk 14]
```diff
@@ -259,7 +262,10 @@ def verify_index_connection(self) -> None:
...
+                            ),
                             dimension: Optional[int] = None,
                             index_params: Optional[dict] = None
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:270]: I think this was replaced by `spec`, right? (for example, `pod_type` used to go in here, but now it will be part of `spec` in case of non-serverless index).  
We can probably delete this argument.

File: src/canopy/knowledge_base/knowledge_base.py [hunk 15]
```diff
@@ -330,13 +321,12 @@ def create_canopy_index(self,
...
-                             'indexed': indexed_fields
-                         },
-                         timeout=TIMEOUT_INDEX_CREATE,
-                         **index_params)
+            self._pinecone_client.create_index(
+                name=self.index_name,
+                dimension=dimension,
+                spec=spec,
+                timeout=TIMEOUT_INDEX_CREATE,
+                **index_params)
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:329]: Again, check me if I'm wrong but I think this is redundant given that all the params that use to go here are now part of `spec`

File: tests/e2e/test_app.py [hunk 16]
```diff
@@ -52,9 +53,16 @@ def index_name(testrun_uid):
...
     kb = KnowledgeBase(index_name=index_name)
-    kb.create_canopy_index(indexed_fields=["test"])
+
+    # System and E2E tests are running in parallel and try to create
+    # indexes at the same time.
+    # DB raises an exception when we create two indexes at the same time.
+    # In order to avoid the exception, we create the index here five seconds
+    # later than the system tests.
+    # TODO: Remove the sleep after the DB is fixed.
+    time.sleep(5)
```
Comment (igiloh-pinecone) [inline tests/e2e/test_app.py:64]: I think it's better to wrap this with a `@retry(stop_after_attempt=1, wait_random=5)` then a naive `sleep`.  
The `sleep` will only work for a single `pytest` run on the same machine - but in reality we run CI workflows in a matrix, so many jobs are running in parallel.

File: tests/system/knowledge_base/test_knowledge_base.py [hunk 17]
```diff
@@ -221,12 +217,11 @@ def encoded_chunks(documents, chunker, encoder):
...
 def test_create_index(index_full_name, knowledge_base):
     assert knowledge_base.index_name == index_full_name
-    assert index_full_name in pinecone.list_indexes()
-    assert index_full_name == index_full_name
+    assert index_full_name in list_canopy_indexes()
     assert knowledge_base._index.describe_index_stats()
 
 
-def test_list_indexes(index_full_name):
+def test_list_indexes(knowledge_base, index_full_name):
```
Comment (igiloh-pinecone) [inline tests/system/knowledge_base/test_knowledge_base.py:224]: Redundant

File: tests/system/knowledge_base/test_knowledge_base.py [hunk 18]
```diff
@@ -466,18 +459,6 @@ def test_connect_after_delete(knowledge_base):
     assert "does not exist or was deleted" in str(e.value)
 
 
-def test_create_with_text_in_indexed_field_raise(index_name,
```
Comment (igiloh-pinecone) [inline tests/system/knowledge_base/test_knowledge_base.py:469]: Please don't delete this test but instead mark it with `@pytest.mark.skip()`.  
We would still need it for non-serverless index testing

**Review by igiloh-pinecone — COMMENTED** [review]
@izellevy All in all LGTM!

Please see a few suggestions

File: src/canopy/knowledge_base/knowledge_base.py (hunk 12 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:108]: Moved it 👍

File: src/canopy/knowledge_base/knowledge_base.py (hunk 15 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:329]: metric and timeout are not part of spec, so worth keeping I guess

File: .github/workflows/release.yml (hunk 9 above)
Comment (izellevy) [reply .github/workflows/release.yml:23]: Good point fixed!

File: src/canopy/knowledge_base/knowledge_base.py (hunk 14 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:270]: still timeout and metric can be in index_params

File: src/canopy/knowledge_base/knowledge_base.py (hunk 13 above)
Comment (igiloh-pinecone) [reply src/canopy/knowledge_base/knowledge_base.py:267]: Oh, sorry, I was missing the point that this is part of the function definition...

In this case, the pythonic way would be `sepc: [Optional[Union[Dict, ServerlessSpec, PodSpec]] = None`, and then inside `__init__()` define set this default spec you crated here.

File: src/canopy/knowledge_base/knowledge_base.py (hunk 14 above)
Comment (igiloh-pinecone) [reply src/canopy/knowledge_base/knowledge_base.py:270]: So let's just split them to `metric: str = 'cosine', timeout: Optional[int]...` and that's it.  
I don't see a point keeping this arbitrary `index_params` dict which is different from the `spec`, it's very confusing.

File: README.md [hunk 19]
```diff
@@ -1,10 +1,10 @@
 # Canopy
 
 <p align="center">
-<a href="https://pypi.org/project/fastapi" target="_blank">
+<a href="https://pypi.org/project/canopy-sdk" target="_blank">
```
Comment (igiloh-pinecone) [inline README.md:4]: @izellevy this change is actually not related to this PR. Can you please remove it, and @miararoy will PR this straight to `main`?

Comment (izellevy) [reply README.md:4]: Sure

**Review by igiloh-pinecone — APPROVED** [review]
LGTM!
